lib/media/timestamp_offset_corrector.js

```diff
@@ -0,0 +1,194 @@
+     */
+    this.contentTypeMediaData_ = new Map();
+
+    /** @private {?shaka.extern.StreamingConfiguration} */
+    this.config_ = null;
+
+    /** @private {?function(!Event)} */
+    this.onEvent_ = onEvent;
+  }
+
+  /**
+   * Configure timestamp offset corrector
+   *
+   * @param {shaka.extern.StreamingConfiguration} config
+   */
+  configure(config) {
+    this.config_ = config;
+  }
+
+  /**
+   * Returns MediaData for a content type.
+   *
+   * @param {shaka.util.ManifestParserUtils.ContentType} contentType
+   * @return {!shaka.media.TimestampOffsetCorrector.MediaData_}
+   * @private
+   */
+  getMediaData_(contentType) {
+    if (this.contentTypeMediaData_.has(contentType)) {
+      return this.contentTypeMediaData_.get(contentType);
+    }
+    /** @type {Map<number,number>} */
+    const trackTimescales = new Map();
+
+    /** @type {shaka.media.TimestampOffsetCorrector.MediaData_} */
+    const mediaData = {
+      timestampOffset: null,
+      originalTimestampOffset: null,
+      trackTimescales: trackTimescales,
+    };
+    this.contentTypeMediaData_.set(contentType, mediaData);
+    return mediaData;
+  }
+
+  /**
+   * Applies corrected timestamp offset to a segment reference
+   *
+   * @param {shaka.util.ManifestParserUtils.ContentType} contentType
+   *   The contentType of the segment reference e.g. 'audio', 'video'
+   * @param {shaka.media.SegmentReference} reference
+   *   A media segment reference
+   */
+  correctTimestampOffset(contentType, reference) {
+    const mediaData = this.getMediaData_(contentType);
+
+    // If the original timestampOffset has not changed since we
+    // last determined the correct timestamp offset then
+    // correct the timestamp offset in the segment reference.
+    if (mediaData.timestampOffset !== null &&
+        mediaData.originalTimestampOffset !== null &&
+        reference.timestampOffset === mediaData.originalTimestampOffset) {
+      reference.timestampOffset = mediaData.timestampOffset;
+    }
+  }
+
+
+  /**
+   * Detects discrepancies in timestamp offset and saves that info for
+   * use in correctTimestampOffset().
+   *
+   * @param {shaka.util.ManifestParserUtils.ContentType} contentType
+   *   The contentType of the segment reference e.g. 'audio', 'video'
+   * @param {shaka.media.SegmentReference} reference
+   *   A media segment reference
+   * @param {BufferSource} segmentData
+   *   The data for a media segment
+   * @return {boolean} true if the timestampOffset has not already been
+   *   corrected and we detect a discrepancy in the timestampOffset.
+   */
+  checkAndCorrectTimestampOffset(contentType, reference, segmentData) {
+    const mediaData = this.getMediaData_(contentType);
+
+    // If timestampOffset has already been corrected do nothing.
+    if (mediaData.timestampOffset !== null &&
+        mediaData.timestampOffset === reference.timestampOffset) {
+      return false;
+    }
+
+    const segmentStartTime =
```

Comment:
If I'm not mistaken, this comes out to 0 for all non-MP4 content, which would trigger a large correction for each segment.

Can you please skip this new functionality for other content types?  I would hate to see this break all WebM content, for example.